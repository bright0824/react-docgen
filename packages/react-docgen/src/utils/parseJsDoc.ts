import doctrine from 'doctrine';
import type { Type, Annotation, Tag } from 'doctrine';

type ExcludesNullish = <T>(x: T | 0 | false | null | undefined) => x is T;
type JsDocType = JsDocBaseType | JsDocElementsType;

interface JsDocBaseType {
  name: string;
}

interface JsDocElementsType extends JsDocBaseType {
  elements: JsDocType[];
}

interface JsDocProperty {
  description: string | null;
  type: JsDocType | null;
}
interface JsDocParam extends JsDocProperty {
  name: string;
  optional?: boolean;
}

interface JsDoc {
  description: string | null;
  params: JsDocParam[];
  returns: JsDocProperty | null;
}

function getType(tagType: Type | null | undefined): JsDocType | null {
  if (!tagType) {
    return null;
  }

  switch (tagType.type) {
    case 'NameExpression':
      // {a}
      return { name: tagType.name };
    case 'UnionType':
      // {a|b}
      return {
        name: 'union',
        elements: tagType.elements
          .map(element => getType(element))
          .filter(Boolean as unknown as ExcludesNullish),
      };
    case 'AllLiteral':
      // {*}
      return { name: 'mixed' };
    case 'TypeApplication':
      // {Array<string>} or {string[]}
      return {
        name: 'name' in tagType.expression ? tagType.expression.name : '',
        elements: tagType.applications
          .map(element => getType(element))
          .filter(Boolean as unknown as ExcludesNullish),
      };
    case 'ArrayType':
      // {[number, string]}
      return {
        name: 'tuple',
        elements: tagType.elements
          .map(element => getType(element))
          .filter(Boolean as unknown as ExcludesNullish),
      };
    default: {
      const typeName =
        'name' in tagType && tagType.name
          ? tagType.name
          : 'expression' in tagType &&
            tagType.expression &&
            'name' in tagType.expression
          ? tagType.expression.name
          : null;

      if (typeName) {
        return { name: typeName };
      } else {
        return null;
      }
    }
  }
}

function getOptional(tag: Tag): boolean {
  return !!(tag.type && tag.type.type && tag.type.type === 'OptionalType');
}

// Add jsdoc @return description.
function getReturnsJsDoc(jsDoc: Annotation): JsDocProperty | null {
  const returnTag = jsDoc.tags.find(
    tag => tag.title === 'return' || tag.title === 'returns',
  );

  if (returnTag) {
    return {
      description: returnTag.description,
      type: getType(returnTag.type),
    };
  }

  return null;
}

// Add jsdoc @param descriptions.
function getParamsJsDoc(jsDoc: Annotation): JsDocParam[] {
  if (!jsDoc.tags) {
    return [];
  }

  return jsDoc.tags
    .filter(tag => tag.title === 'param')
    .map(tag => {
      return {
        name: tag.name || '',
        description: tag.description,
        type: getType(tag.type),
        optional: getOptional(tag),
      };
    });
}

export default function parseJsDoc(docblock: string): JsDoc {
  const jsDoc = doctrine.parse(docblock);

  return {
    description: jsDoc.description || null,
    params: getParamsJsDoc(jsDoc),
    returns: getReturnsJsDoc(jsDoc),
  };
}
