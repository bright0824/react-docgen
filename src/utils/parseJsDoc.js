/*
 * Copyright (c) 2015, Facebook, Inc.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @flow
 */

import doctrine from 'doctrine';

type JsDoc = {
  description: ?string;
  params: [{
    name: string;
    description: ?string;
    type: ?{name: string};
    optional?: boolean;
  }];
  returns: ?{
    description: ?string;
    type: ?{name: string};
  };
};

function getType(tag) {
  if (!tag.type) {
    return null;
  }
  return {name: tag.type.name ? tag.type.name : tag.type.expression.name};
}

function getOptional(tag) {
  if (tag.type && tag.type.type && tag.type.type === 'OptionalType') {
    return true;
  }
  return;
}

// Add jsdoc @return description.
function getReturnsJsDoc(jsDoc) {
  const returnTag = jsDoc.tags.find(
    tag => tag.title === 'return' || tag.title === 'returns'
  );
  if (returnTag) {
    return {
      description: returnTag.description,
      type: getType(returnTag),
    };
  }
  return null;
}

// Add jsdoc @param descriptions.
function getParamsJsDoc(jsDoc) {
  if (!jsDoc.tags) {
    return [];
  }
  return jsDoc.tags
    .filter(tag => tag.title === 'param')
    .map(tag => {
      return {
        name: tag.name,
        description: tag.description,
        type: getType(tag),
        optional: getOptional(tag),
      };
    });
}

export default function parseJsDoc(docblock: string): JsDoc {
  const jsDoc = doctrine.parse(docblock);

  return {
    description: jsDoc.description || null,
    params: getParamsJsDoc(jsDoc),
    returns: getReturnsJsDoc(jsDoc),
  };
}
