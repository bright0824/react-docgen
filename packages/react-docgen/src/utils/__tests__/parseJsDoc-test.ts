import parseJsDoc from '../parseJsDoc.js';
import { describe, expect, test } from 'vitest';

describe('parseJsDoc', () => {
  describe('description', () => {
    test('extracts the method description in jsdoc', () => {
      const docblock = `
        Don't use this!
      `;

      expect(parseJsDoc(docblock)).toMatchSnapshot();
    });
  });
  describe('@param', () => {
    const docBlocks = {
      'extracts jsdoc description': '@param bar test',
      'extracts jsdoc empty description': '@param {string} bar',
      'extracts jsdoc union type param': '@param {string|Object|some[]} bar',
      'extracts jsdoc optional': '@param {string=} bar',
      'extracts jsdoc typed array': '@param {[string, number]} bar',
    };

    Object.keys(docBlocks).forEach((name) => {
      const docBlock = docBlocks[name];

      test(name, () => {
        expect(parseJsDoc(docBlock)).toMatchSnapshot();
      });
    });
  });
  describe('@returns', () => {
    const docBlocks = {
      'extracts jsdoc types': '@returns {string}',
      'extracts jsdoc mixed types': '@returns {*}',
      'extracts description from jsdoc': '@returns The number',
      'works with @return': '@return The number',
      'extracts jsdoc typed array': '@param {[string, number]} bar',
    };

    Object.keys(docBlocks).forEach((name) => {
      const docBlock = docBlocks[name];

      test(name, () => {
        expect(parseJsDoc(docBlock)).toMatchSnapshot();
      });
    });
  });
});
