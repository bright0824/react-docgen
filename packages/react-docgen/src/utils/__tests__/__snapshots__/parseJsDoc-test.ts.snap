// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`parseJsDoc > @param > extracts jsdoc description 1`] = `
{
  "description": null,
  "params": [
    {
      "description": "test",
      "name": "bar",
      "optional": false,
      "type": null,
    },
  ],
  "returns": null,
}
`;

exports[`parseJsDoc > @param > extracts jsdoc empty description 1`] = `
{
  "description": null,
  "params": [
    {
      "description": null,
      "name": "bar",
      "optional": false,
      "type": {
        "name": "string",
      },
    },
  ],
  "returns": null,
}
`;

exports[`parseJsDoc > @param > extracts jsdoc optional 1`] = `
{
  "description": null,
  "params": [
    {
      "description": null,
      "name": "bar",
      "optional": true,
      "type": {
        "name": "string",
      },
    },
  ],
  "returns": null,
}
`;

exports[`parseJsDoc > @param > extracts jsdoc typed array 1`] = `
{
  "description": null,
  "params": [
    {
      "description": null,
      "name": "bar",
      "optional": false,
      "type": {
        "elements": [
          {
            "name": "string",
          },
          {
            "name": "number",
          },
        ],
        "name": "tuple",
      },
    },
  ],
  "returns": null,
}
`;

exports[`parseJsDoc > @param > extracts jsdoc union type param 1`] = `
{
  "description": null,
  "params": [
    {
      "description": null,
      "name": "bar",
      "optional": false,
      "type": {
        "elements": [
          {
            "name": "string",
          },
          {
            "name": "Object",
          },
          {
            "elements": [
              {
                "name": "some",
              },
            ],
            "name": "Array",
          },
        ],
        "name": "union",
      },
    },
  ],
  "returns": null,
}
`;

exports[`parseJsDoc > @returns > extracts description from jsdoc 1`] = `
{
  "description": null,
  "params": [],
  "returns": {
    "description": "The number",
    "type": null,
  },
}
`;

exports[`parseJsDoc > @returns > extracts jsdoc mixed types 1`] = `
{
  "description": null,
  "params": [],
  "returns": {
    "description": null,
    "type": {
      "name": "mixed",
    },
  },
}
`;

exports[`parseJsDoc > @returns > extracts jsdoc typed array 1`] = `
{
  "description": null,
  "params": [
    {
      "description": null,
      "name": "bar",
      "optional": false,
      "type": {
        "elements": [
          {
            "name": "string",
          },
          {
            "name": "number",
          },
        ],
        "name": "tuple",
      },
    },
  ],
  "returns": null,
}
`;

exports[`parseJsDoc > @returns > extracts jsdoc types 1`] = `
{
  "description": null,
  "params": [],
  "returns": {
    "description": null,
    "type": {
      "name": "string",
    },
  },
}
`;

exports[`parseJsDoc > @returns > works with @return 1`] = `
{
  "description": null,
  "params": [],
  "returns": {
    "description": "The number",
    "type": null,
  },
}
`;

exports[`parseJsDoc > description > extracts the method description in jsdoc 1`] = `
{
  "description": "Don't use this!",
  "params": [],
  "returns": null,
}
`;
